yarn-project/simulator/src/avm/avm_tree.ts

```diff
@@ -0,0 +1,808 @@
+   * @param index - The index of the leaf that the sibling path is derived from
+   * @param siblingPath - The sibling path of the index
+   */
+  insertSiblingPath(index: bigint, siblingPath: Fr[]): void {
+    for (let i = 0; i < siblingPath.length; i++) {
+      // Flip(XOR) the last bit because we are inserting siblings of the leaf
+      const sibIndex = index ^ 1n;
+      this.updateLeaf(siblingPath[i], sibIndex, this.depth - i);
+      index >>= 1n;
+    }
+  }
+
+  /**
+   * This is a helper function that computes the frontier leaf slots of a tree from the leaf count
+   * @param leafCount - The number of leaves in the tree
+   */
+  // Do we really need LeafCount to be a bigint - log2 is on numbers only
+  static computeFrontierLeafSlots(leafCount: number) {
+    const numFrontierEntries = Math.floor(Math.log2(leafCount)) + 1;
+    const frontierIndices = [];
+    for (let i = 0; i < numFrontierEntries; i++) {
+      if (leafCount === 0) {
+        frontierIndices.push(0);
+      } else if (leafCount % 2 === 0) {
+        frontierIndices.push(leafCount - 2);
+      } else {
+        frontierIndices.push(leafCount - 1);
+      }
+      leafCount >>= 1;
+    }
+    return frontierIndices;
+  }
+
+  /**
+   * This derives the frontier and inserts them into the tree
+   * @param treeDb - The treeDb to be queried for sibling paths
+   * @param merkleId - The treeId of the tree to be queried for sibling paths
+   */
+  async initializeFrontier(treeDb: MerkleTreeReadOperations, merkleId: MerkleTreeId): Promise<void> {
+    // The frontier indices are sorted from the leaf to root
+    const frontierIndices = EphemeralAvmTree.computeFrontierLeafSlots(Number(this.leafCount));
+    // The frontier indices are level-based - i.e. index N at level L.
+    // Since we can only ask the DB for paths from the root to the leaf, we do the following complicated calculations
+    // 1) The goal is to insert the frontier node N at level L into the tree.
+    // 2) We get the path to a leaf that passes through the frontier node we want (there are multiple paths so we just pick one)
+    // 3) We can only get sibling paths from the root to the leaf, so we get the sibling path of the leaf from (2)
+    // NOTE: This is terribly inefficient and we should probably change the DB API to allow for getting paths to a node
+
+    // These are leaf indexes that pass through the frontier nodes
+    const siblingValue = [];
+    for (let i = 0; i < frontierIndices.length; i++) {
+      // The path we need has the frontier node as a sibling
+      const index = BigInt(frontierIndices[i] ^ 1) << BigInt(i);
+      const path = await treeDb.getSiblingPath(merkleId, index);
+
+      const frontierPath = this._derivePathLE(BigInt(frontierIndices[i]), this.depth - i);
+      siblingValue.push(path.toFields()[i]);
+      this.tree = this._insertLeaf(siblingValue[i], frontierPath, this.depth - i, this.tree, true);
+    }
+
+    this.frontier = siblingValue;
+  }
```

Comment:
I don't know what a frontier is, and tbh I don't understand why we need to do insertions here on initialization of the tree. Is this grabbing the rightmost leaf from the DB and inserting its entire path of nodes into the ephemeral one?

Reply:
Not quite the rightmost leaf, the frontier essentially contains the minimum number of nodes (one from each level of the tree) required to be able to append a new element to the tree.

Since the tree will be empty at the start (since all the information is in the DB), we need to "seed" the tree for the first time so it nows how to handle it's first append. The nodes to do this come from the frontier - so we derive it from the DB